contracts/LockedNORIV2.sol

```diff
@@ -0,0 +1,791 @@
+   * ERC-1820 registry
+   *
+   * @dev Registering that LockedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  /**
+   * @notice A mapping from grantee to grant
+   */
+  mapping(address => TokenGrant) private _grants;
+
+  /**
+   * @notice The BridgedPolygonNORI contract that this contract wraps tokens for
+   */
+  IERC20Upgradeable private _bridgedPolygonNori;
+
+  /**
+   * @notice The [ERC-1820](https://eips.ethereum.org/EIPS/eip-1820) pseudo-introspection registry
+   * contract
+   *
+   * @dev Registering that LockedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  IERC1820RegistryUpgradeable private _erc1820;
+
+  /**
+   * @notice Emitted on successful creation of a new grant.
+   */
+  event TokenGrantCreated(
+    address indexed recipient,
+    uint256 indexed amount,
+    uint256 indexed startTime,
+    uint256 vestEndTime,
+    uint256 unlockEndTime
+  );
+
+  /**
+   * @notice Emitted on when the vesting portion of an active grant is terminated.
+   */
+  event UnvestedTokensRevoked(
+    uint256 indexed atTime,
+    address indexed from,
+    uint256 indexed quantity
+  );
+
+  /**
+   * @notice Emitted on withdwal of fully unlocked tokens.
+   */
+  event TokensClaimed(
+    address indexed from,
+    address indexed to,
+    uint256 quantity
+  );
+
+  /**
+   * @notice Emitted when the underlying token contract address is updated due to migration.
+   */
+  event UnderlyingTokenAddressUpdated(address from, address to);
+
+  /**
+   * @dev The tokensReceived callback reverts so we avoid receipt of ERC777 tokens.
+   */
+  function tokensReceived(
+    address,
+    address,
+    address,
+    uint256,
+    bytes calldata,
+    bytes calldata
+  ) external override {
+    revert("Funding via tokensReceived callback in not supported");
+  }
+
+  /**
+   * @notice Wraps minting of wrapper token and grant setup.
+   *
+   * @dev If `startTime` is zero no grant is set up. Satisfies situations where funding of the grant happens over time.
+   *
+   * @param amount uint256 Quantity of `_bridgedPolygonNori` to deposit
+   */
+  function depositFor(address recipient, uint256 amount)
+    external
+    whenNotPaused
+    returns (bool)
+  {
+    require(_grants[recipient].exists, "lNORI: Cannot deposit without a grant");
+    _bridgedPolygonNori.transferFrom(_msgSender(), address(this), amount);
```

Comment:
do we want to use this instead of the permit pattern?

- using this approach: 2 serial transactions (approve + depositFor) (although it might be possible to only ever have to call `approve` once if we give it a high enough balance, erc 20 allows finfinite approvals even if the amount is > the current accounts balance)
- use erc 20 permit: 1 signature, and 1 transactions. requires figuring out typed data signing with fireblocks. Definitely safer, than infinite approvals, and fewer transactions (although signing would feel the same as a transaction in fireblocks since it still requires confirming in the app)

Reply:
I think you'd actually have to also call `createGrant` using the non-permit pattern, right? 3 total transactions I think



`createGrant` -> `approve` -> `depositFor`? Is there a way to simplify that to two?



Permit could be done in 1 signature + 1 tx

Reply:
Yeah definitely intending to revisit this now that we know this upgrade is working cleanly.